### PR TITLE
Realization filter

### PR DESCRIPTION
**Pull request by jorgenherje**
First version of realization filter

Filter with include/exclude and selecting a subset of realization indices for an ensemble. New hook to utilize for retrieving realization indices for an ensemble. If no filtering is applied, the original array is returned.

Filtering on parameter value is to be added soon.

Usage added into `SimulationTimeSeriesMatrix` module for testing.

---

Goal is to start using the new hook for lookup of current set of realization indices for an ensemble. The subset should always be used for calculating statistics.

---

Closes: #330 
partially fixes: #348

### CODE REVIEW

**Review by rubenthoms — CHANGES_REQUESTED** [review]
Great job! A couple of rename suggestions and some minor issues.

File: frontend/src/framework/internal/components/NavBar/rightNavBar.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        if (filterPanelWidth <= 5) {
+            setFilterPanelWidth(15);
+        }
+    }
+
+    function handleFilterPanelCollapseOrExpand() {
+        const newExpanded = !filterPanelExpanded;
+
+        localStorage.setItem("filterPanelExpanded", newExpanded.toString());
+        setFilterPanelExpanded(newExpanded);
```
[the remark this answers is not in the document]
Comment (jorgenherje) [reply frontend/src/framework/internal/components/NavBar/rightNavBar.tsx:33]: That's true!

File: frontend/src/framework/GuiMessageBroker.ts [hunk 2]
```diff
@@ -73,6 +75,8 @@ type GuiStateValueTypes = {
     [GuiState.DataChannelConnectionLayerVisible]: boolean;
     [GuiState.DevToolsVisible]: boolean;
     [GuiState.EditDataChannelConnections]: boolean;
+    [GuiState.FilterPanelWidthInPercent]: number;
+    [GuiState.FilterPanelExpanded]: boolean;
```
[the remark this answers is not in the document]
Comment (jorgenherje) [reply frontend/src/framework/GuiMessageBroker.ts:79]: 👍

File: frontend/src/framework/RealizationFilter.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+export const RealizationFilteringOptionStringMapping = {
+    [RealizationFilteringOption.INCLUDE]: "Include",
+    [RealizationFilteringOption.EXCLUDE]: "Exclude",
+};
+
+export type IndexRangeType = { start: number; end: number };
+export type RealizationIndexSelectionType = IndexRangeType | number;
+
+export class RealizationFilter {
+    private _parentEnsemble: Ensemble;
```
[the remark this answers is not in the document]
Comment (jorgenherje) [reply frontend/src/framework/RealizationFilter.ts:26]: Changed to `_assignedEnsemble` as ensemble is assigned to the filter on creation.

File: frontend/src/framework/components/RealizationPicker/realizationPicker.tsx [hunk 4]
```diff
@@ -146,7 +147,7 @@ const RealizationRangeTag: React.FC<RealizationRangeTagProps> = (props) => {
             />
             <div
                 className={resolveClassNames(
-                    "text-slate-800 hover:text-slate-600 text-sm cursor-pointer flex items-center",
+                    "text-slate-800 hover:text-slate-600 text-sm cursor-pointer flex items-center out",
```
[the remark this answers is not in the document]
Comment (jorgenherje) [reply frontend/src/framework/components/RealizationPicker/realizationPicker.tsx:150]: Have no clue where it comes from, probably a typo.

File: frontend/src/framework/RealizationFilter.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rubenthoms) [reply frontend/src/framework/RealizationFilter.ts:26]: Perfect!

File: frontend/src/framework/internal/components/RealizationFilterSettings/realizationFilterSettings.tsx [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        setRealizationIndexSelections(realizationIndexSelection);
+        setSelectedRangeTags(makeRealizationPickerTagsFromRealizationIndexSelections(realizationIndexSelection));
+        setSelectedFilterType(realizationFilter.getFilterType());
+        setSelectedIncludeOrExcludeFiltering(realizationFilter.getIncludeOrExcludeFilter());
+    }
+
+    function handleSelectedEnsembleChange(newValue: string | undefined) {
+        const ensembleIdent = newValue ? EnsembleIdent.fromString(newValue) : null;
+        if (hasUnsavedChanges) {
+            setCandidateEnsembleIdent(ensembleIdent);
```
Comment (rubenthoms) [inline frontend/src/framework/internal/components/RealizationFilterSettings/realizationFilterSettings.tsx:79]: This is also a duplicate and could be moved to line 78.

**Review by rubenthoms — APPROVED** [review]
One small change, otherwise, good to go :+1:

**Pull request merged**